### PR TITLE
ARROW-13052: [Gandiva][C++] Add regexp_extract function

### PR DESCRIPTION
**Pull request by Johnnathanalmeida**
Implements the REGEXP_EXTRACT function based on [the Hive implementation](https://www.revisitclass.com/hadoop/regexp_extract-function-in-hive-with-examples/).

### CODE REVIEW

**Comment by Johnnathanalmeida** [comment]
@github-actions crossbow submit java-jars

File: cpp/src/gandiva/tests/projector_test.cc [hunk 1]
```diff
@@ -2624,4 +2624,54 @@ TEST_F(TestProjector, TestNextDay) {
...
+  auto extract_expr = TreeExprBuilder::MakeExpression(regexp_extract_func, field_extract);
+
+  std::shared_ptr<Projector> projector;
+  auto status = Projector::Make(schema, {extract_expr}, TestConfiguration(), &projector);
+  EXPECT_TRUE(status.ok()) << status.message();
+
+  // Create a row-batch with some sample data
+  int num_records = 7;
+  auto array0 = MakeArrowArrayUtf8(
+      {"John Doe - 124", "John Doe - 124", "John Doe - 124", "John Doe - 124",
```
[the remark this answers is not in the document]
Comment (Johnnathanalmeida) [reply cpp/src/gandiva/tests/projector_test.cc:2655]: > can you add test with UTF8 entry? e.g. : b路%$大 路%$大a 路%c$大

New tests cases with japanese characters were add

**Review by ViniciusSouzaRoque — APPROVED** [review]
LGFM

**Review by anthonylouisbsb — APPROVED** [review]

**Review by projjal — APPROVED** [review]

**Comment by Johnnathanalmeida** [comment]
@github-actions crossbow submit java-jars

**Pull request closed without merging**